### ISSUE TITLE
Inconsistent format for properties

### ISSUE THREAD

**Issue by snicoll** [spring-cloud/spring-cloud-dataflow-ui#769]
If I deploy a task with no additional customization, I am being presented with a list of properties that are available for that task.

The structure of the properties is not consistent: some are displayed with a full id (prefix + property name) and others are displayed with only their property name. I don't know if that is on purpose but if that is I couldn't figure out why. Also, they don't seem to be ordered.

It looks like that some properties may be missing as well as multiple keys may share the same property name (with different prefixes).

**Comment by sabbyanandan** [comment]
The App specific properties can be [whitelisted](http://docs.spring.io/spring-cloud-dataflow/docs/1.5.0.RELEASE/reference/htmlsingle/#spring-cloud-dataflow-stream-app-whitelisting) and with its [companion metadata artifact](http://docs.spring.io/spring-cloud-dataflow/docs/1.5.0.RELEASE/reference/htmlsingle/#spring-cloud-dataflow-stream-app-metadata-artifact) (via the dedicated plugin), we should only be seeing the subset of properties. However, if we don't whitelist, we end up seeing the entire set of properties available in Boot.

Also, for App specific properties, we recommend the use of "prefixes", so it can be whitelisted and identified correctly. All of our OOTB Apps follow that model _(run: `app info source:http` from the shell)_. With this in place, `app info` in Shell or from the UI, the properties should list correct keys. If it doesn't, it is a bug in the App.

In summary, it is a user responsibility; they'd have to create and whitelist the properties. I'll close this since we cannot do anything more than recommending the companion-artifact approach.

**Comment by snicoll** [comment]
@sabbyanandan that answer doesn't match what I intended to report. Let me try again.

> If I deploy a task with no additional customization

I am not reporting that all the properties are listed by default (though that could be also a valid issue IMO), I am reporting that they are displayed inconsistently.

For instance, there is a property called `init` that maps to `spring.webservices.servlet.init`. How am I supposed to know that? Based on the description?

Going further down I have this: `password` with a description of `Login password`. It could be `spring.elasticsearch.jest.password` or `spring.influx.password` or something else.

Some keys are listed with a full id, others with only the property name.

To be clear, it's not about the properties I may have created for my task (I didn't add any). It is about what is being presented to me by default. Can you please explain why this format is working as designed?

As the focus isn't about whitelisting and creating a proper prefix for custom task properties, I'd like this issue to be reopened. Thanks.

**Comment by snicoll** [comment]
> Feel free to ping us in Gitter (or Slack) for more questions - we typically use GitHub for features/bugs.

Sorry, I’m obviously not doing a very good job of explaining this. Rather than asking a question, I’m reporting a bug with the usability of the UI. I also didn’t mention the `jdbc-source` and, as far as I know, the problem I’m seeing isn’t really related to it. Let me have another go and see if I can explain the problem.

This is what I am using when deploying a simple task with no additional customization. I understand that a metadata file can be provided for whiltelisting properties but it's not mandatory.

<img width="395" alt="pic1" src="https://user-images.githubusercontent.com/490484/40298535-c4eb76f4-5ce3-11e8-9839-b656ebbdbd48.png">

I still don't understand why some properties have a full key and some don't. I haven't made any customization to the metadata so showing those in an inconsistent format looks like a bug to me.

<img width="342" alt="pic3" src="https://user-images.githubusercontent.com/490484/40298641-1414cbf4-5ce4-11e8-9801-f0ad0758c9f8.png">

**Comment by sabbyanandan** [comment]
>  I understand that a metadata file can be provided for whiltelisting properties but it's not mandatory.

It might sound repetitive, but there's nothing more than what we do via whitelisting, is in our control.

As per the design, currently, if the desired properties aren't whitelisted (not just the custom properties; it also applies to every other global Boot properties) or prefixed as needed, we end up showing everything (900+ properties). In your example, `max-interval` - it obviously doesn't apply to your `myjob` Task App; however, it ends up showing on the list because there was nothing whitelisted. So, it is not just about the format, but it is about the applicability of this property to the app itself. 

I attempted to explain it with `jdbc-source` that's why. This App includes custom, common, and as well as it inherits a few Boot properties such as "datasource" k/v pairs. The assumption here is that the developer who's writing the App has knowledge of existing Boot properties and s/he knows how to leverage and prefix them to be used in SCSt/SCT App. It is a big assumption, though. We could improve the App Starters reference guide to clarify it with an example.

That said, there's nothing to do in the UI because the Dashboard lists what's provided by the backend controller. If we do some filtering in UI alone, the Shell and REST tools will have different behavior - we have to consistent because we have different user personas using different client tools.

Also, we were recently discussing the option of disabling no properties from the display, if the App doesn't whitelist properties. Showing 900+ unrelated properties can be argued as bad UX. If we do the filtering, it will be done in the backend, so the shell, ui, and rest can produce a consistent behavior.

I'd like to request @artembilan, @soby, or @markpollack who can help explain this differently.

**Comment by soby** [comment]
wrong @soby maybe?

**Comment by sabbyanandan** [comment]
cc:/ @sobychacko

**Comment by snicoll** [comment]
@sabbyanandan it might also sound repetitive but I am solely reporting on the inconsistent format for properties here and, as far as I can tell, I haven't got an explanation as why some properties are displayed with only their names and some with their full ids. Surely, this is in your control.

For an example, see second screenshot (`spring.rabbitmq.listener.simple.retry.enabled` vs. `enabled`).

I am happy to create another issue in whatever issue tracker you point me at to the problem of listing all properties by default and/or not making whitelisting mandatory (my gut feeling is that some opinionated default exclude would probably be the best compromise).

**Comment by artembilan** [comment]
@sabbyanandan ,

I think what @snicoll  requests is really valid end-user experience with our UI:

1. We need to order properties for consistency when we reload the page.
2. I think the best feeling from end-user perspective is like to see properties grouped by the same prefix. This way we cover both whitelisted and non-whitelisted cases. That is already up to end-user to use full name or short one. Personally I prefer the full name because sometimes in testing I fail to remember when whitelisting trick is available. 😄 
3. Of course, we still lack of description in some properties, but that's already a different story for each app starter...

Hope that helps a bit...

**Comment by sabbyanandan** [comment]
Thanks, @artembilan. I don't think ordering is what Stephane is referring to as a problem here. His feedback relates to the properties with and without prefixes, and the applicability of them from a developer's pov. Other than be able to whitelist and an explicit prefix _included by the developer_, I do not know how else we would align and control the properties. 

That said, I agree the differences between `spring.rabbitmq.listener.simple.retry.enabled` vs. `enabled` could be opaque to the end users. I'd be interested in ideas to help educate them on this regard.

Let's see if @sobychacko or @markpollack can chime-in. I would move this issue to App Starters backlog once we hear from them.

**Comment by dturanski** [comment]
Did some digging into this.  Looking at the json returned from the `http://localhost:9393/task/<name>` I see things like  
```
{
      "id": "spring.rabbitmq.cache.channel.checkout-timeout",
      "name": "checkout-timeout",
...
}
```

and in many cases id and name are the same:

```
{
      "id": "spring.data.cassandra.repositories.type",
      "name": "spring.data.cassandra.repositories.type",
...
}
```
I haven't looked at how this content is produced and why the inconsistency, but it seems like displaying `id` instead of `name` would fix it.

**Comment by dturanski** [comment]
Looking into it further we should be showing `id` or possibly group by prefix included in the id.

**Comment by jvalkeal** [comment]
@snicoll @dturanski Happened to see this issue and regarding `spring.rabbitmq.listener.simple.max-attempts` and its inconsistent appearance of group id, I know what's going on as I fixed some issues related to this this week. Lets look two metadata groups boot have, `spring.rabbitmq.listener.direct` and `spring.rabbitmq.listener.simple` as shown below:

```
{
  "name": "spring.rabbitmq.listener.direct",
  "type": "org.springframework.boot.autoconfigure.amqp.RabbitProperties$DirectContainer",
  "sourceType": "org.springframework.boot.autoconfigure.amqp.RabbitProperties$Listener",
  "sourceMethod": "getDirect()"
},


{
  "name": "spring.rabbitmq.listener.simple",
  "type": "org.springframework.boot.autoconfigure.amqp.RabbitProperties$SimpleContainer",
  "sourceType": "org.springframework.boot.autoconfigure.amqp.RabbitProperties$Listener",
  "sourceMethod": "getSimple()"
},
```

When you i.e. use `ConfigurationMetadataRepositoryJsonBuilder` manually to build repository to get access to `ConfigurationMetadataProperty`'s, `RawConfigurationMetadata` is getting slightly confused when it's trying to strip group id from a fully qualified property name to set a simple name. See below:

https://github.com/spring-projects/spring-boot/blob/ddc8b6f6b9fa8e8c82e0fb843f961d0c76d3167d/spring-boot-project/spring-boot-tools/spring-boot-configuration-metadata/src/main/java/org/springframework/boot/configurationmetadata/RawConfigurationMetadata.java#L73-L79


If two different properties have same `sourceType`, then `RawConfigurationMetadata.getSource(String)` always return null for atleast one of them, and later in `RawConfigurationMetadata.resolveName(ConfigurationMetadataItem)` it doesn't strip group id to re-set a property name.

We saw this in `spring-cloud-deployer-kubernetes` and decided to use two different classes instead of one base class for two nested groups. https://github.com/spring-cloud/spring-cloud-deployer-kubernetes/commit/58f150739094c81fe03360e7a74d8393cc352450

Not sure if boot could also check `sourceMethod` if those are different when trying to find matching `ConfigurationMetadataSource`.

**Comment by jvalkeal** [comment]
No hold on, I remembered this wrong. It wasn't because `null` is returned but because `RawConfigurationMetadata.getSource(String)` is returning wrong `ConfigurationMetadataSource` and then later group id doesn't match so group id is not stripped.
```
if (hasLength(groupId) && id.startsWith(dottedPrefix)) {
	String name = id.substring(dottedPrefix.length());
	item.setName(name);
}
```

**Comment by jvalkeal** [comment]
Closing as original issue with UI were fixed.
